adapters/epom/epom.go

```diff
@@ -0,0 +1,153 @@
+package epom
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/mxmCherry/openrtb"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+type adapter struct {
+	endpoint string
+}
+
+// Builder builds a new instance of the Epom adapter for the given bidder with the given config.
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter) (adapters.Bidder, error) {
+	bidder := &adapter{
+		endpoint: config.Endpoint,
+	}
+	return bidder, nil
+}
+
+func (a *adapter) MakeRequests(request *openrtb.BidRequest, reqInfo *adapters.ExtraRequestInfo) (requests []*adapters.RequestData, errors []error) {
+	rq, errs := a.makeRequest(request)
+
+	if len(errs) > 0 {
+		return nil, errs
+	}
+
+	if rq != nil {
+		requests = append(requests, rq)
+	}
+
+	return requests, nil
+}
+
+func (a *adapter) makeRequest(request *openrtb.BidRequest) (*adapters.RequestData, []error) {
+	var errs []error
+	var validImps []openrtb.Imp
+
+	if request.Device == nil || request.Device.IP == "" {
+		return nil, []error{&errortypes.BadInput{
+			Message: "ipv4 address is required field",
+		}}
+	}
+
+	for _, imp := range request.Imp {
+		if err := preprocess(&imp); err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		validImps = append(validImps, imp)
+	}
+
+	if len(validImps) == 0 {
+		return nil, errs
+	}
+
+	request.Imp = validImps
+
+	reqJSON, err := json.Marshal(request)
+	if err != nil {
+		errs = append(errs, err)
+		return nil, errs
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	return &adapters.RequestData{
+		Method:  "POST",
+		Uri:     a.endpoint,
+		Body:    reqJSON,
+		Headers: headers,
+	}, errs
+}
+
+func preprocess(imp *openrtb.Imp) error {
+	if imp.Audio != nil {
+		return &errortypes.BadInput{
+			Message: "Audio is not supported",
+		}
+	}
```

Comment:
This check is also covered by PBS-Core, since audio excluded as an acceptable type in the yaml file. You can remove the entire `preprocess` method, it's tests, and the `for _, imp := range request.Imp {` loop in ` MakeRequests`. The only check you need to keep is for the ipv4.

In the end, I think this would be sufficient:
```
func (a *adapter) makeRequest(request *openrtb.BidRequest) (*adapters.RequestData, []error) {
	if request.Device == nil || request.Device.IP == "" {
		return nil, []error{&errortypes.BadInput{
			Message: "ipv4 address is required field",
		}}
	}

	reqJSON, err := json.Marshal(request)
	if err != nil {
		return nil, []error{err}
	}

	headers := http.Header{}
	headers.Add("Content-Type", "application/json;charset=utf-8")
	headers.Add("Accept", "application/json")
	return &adapters.RequestData{
		Method:  "POST",
		Uri:     a.endpoint,
		Body:    reqJSON,
		Headers: headers,
	}, nil
}
```

Reply:
Thank you! Used your code. Check please.